### PR TITLE
Better error unknown

### PR DESCRIPTION
**Pull request by kevmoo**

### CODE REVIEW

File: lib/src/json_serializable_generator.dart [hunk 1]
```diff
@@ -58,6 +58,17 @@ class JsonSerializableGenerator
     // TODO: support overriding the field set with an annotation option
     var fieldsList = classElement.fields.where((e) => !e.isStatic).toList();
 
+    var undefinedFields =
+        fieldsList.where((fe) => fe.type.isUndefined).toList();
```
Comment (natebosch) [inline lib/src/json_serializable_generator.dart]: why do you need toList?

File: lib/src/json_serializable_generator.dart [hunk 2]
```diff
@@ -151,6 +162,19 @@ class JsonSerializableGenerator
       fieldsToSet.remove(arg.name);
     }
 
+    var undefinedArgs = [ctorArguments, ctorNamedArguments]
+        .expand((l) => l)
```
Comment (natebosch) [inline lib/src/json_serializable_generator.dart]: we need Iterable.followedBy :)

File: lib/src/json_serializable_generator.dart [hunk 3]
```diff
@@ -151,6 +162,19 @@ class JsonSerializableGenerator
...
+        .where((pe) => pe.type.isUndefined)
+        .toList();
```
Comment (natebosch) [inline lib/src/json_serializable_generator.dart]: why toList()?

File: lib/src/json_serializable_generator.dart [hunk 4]
```diff
@@ -296,15 +320,17 @@ class JsonSerializableGenerator
       {ParameterElement ctorParam}) {
     name = _fieldToJsonMapKey(name, field);
     var result = "json['$name']";
-    return _writeAccessToJsonValue(result, field.type, ctorParam: ctorParam);
+    DartType targetType;
```
Comment (natebosch) [inline lib/src/json_serializable_generator.dart]: [optional] `var targetType = ctorParam?.type ?? field.type;` or `var targetType = (ctorParam ?? field).type;`

File: lib/src/json_serializable_generator.dart (hunk 2 above)
Comment (kevmoo) [inline lib/src/json_serializable_generator.dart]: Indeed!

File: lib/src/json_serializable_generator.dart (hunk 1 above)
Comment (kevmoo) [inline lib/src/json_serializable_generator.dart]: I hate reiterating iterables – even trivially. Can have unintended perf and behavior affects.

So I just never do it.

File: lib/src/json_serializable_generator.dart (hunk 3 above)
Comment (kevmoo) [inline lib/src/json_serializable_generator.dart]: ditto

File: lib/src/json_serializable_generator.dart (hunk 4 above)
Comment (kevmoo) [inline lib/src/json_serializable_generator.dart]: nice

**Pull request merged**